### PR TITLE
Feature: Floating turbine Cp/Ct/Tilt support

### PR DESCRIPTION
**Pull request by bayc**
**Feature or improvement description**
This PR adds support for floating turbines in the form of Cp/Ct curve corrections for tilt of the floating platform based on wind speed.

**Related issue, if one exists**
#515
#517 
#526 
#527 

**Impacted areas of the software**
A lot of the simulation code, focusing mostly in `turbine.py` and `farm.py`. Updates to `floris_interface.py` as well as the pre-defined turbines.

**Additional supporting information**
Specific tasks involved in this PR:

- [x] Add reference tilt to inputs and `turbine.py`
- [x] Add tilt/wind speed table to inputs and `turbine.py`
- [x] Build out tilt values in `farm.py`
- [x] Update solvers to support tilt values/corrections
- [x] Re-scope power function in `turbine.py`
- [x] Add tilt angle support to `floris_interface.py`
- [x] Update all predefined turbine definitions to include reference tilt
- [x] Update unit tests
- [x] Update regression tests

**Test results, if applicable**
Currently all tests pass. Unit tests were added for the new/rescoped functions in `turbine.py`, while other unit tests were updated appropriately for tilt inputs. Note the regression test baseline values did not change. This was intentional to show that the tilt correction code does not cause a change in calculated values when the tilt of the turbine is equal to the reference tilt of that turbine.

### CODE REVIEW

**Comment by paulf81** [comment]
Yes @bayc I think you've captured it all nicely, this looks good!

File: .gitignore [hunk 1]
```diff
@@ -32,3 +32,9 @@ examples/SLSQP.out
 
 # Log files
 *.log
+
+# User-specific files
```
Comment (rafmudaf) [inline .gitignore:36]: Are these relevant to keep or left over from local changes?

File: examples/inputs_floating/gch_fixed.yaml [hunk 2]
```diff
@@ -0,0 +1,226 @@
+
```
Comment (rafmudaf) [inline examples/inputs_floating/gch_fixed.yaml:1]: The gch.yaml input file also serves as the reference for the documentation. It might be worthwhile to remove all of the comments here for readability.
Same comment for the group of input files added here.

File: floris/simulation/turbine.py [hunk 3]
```diff
@@ -319,6 +489,38 @@ def __attrs_post_init__(self) -> None:
...
+
+        if any(el.ndim > 1 for el in inputs):
+            raise ValueError("tilt and wind_speed inputs must be 1-D.")
+
+        if len({self.tilt.size, self.wind_speeds.size}) > 1:
+            raise ValueError("tilt and wind_speed tables must be the same size.")
+
+        # Remove any duplicate wind speed entries
+        _, duplicate_filter = np.unique(self.wind_speeds, return_index=True)
+        object.__setattr__(self, "tilt", self.tilt[duplicate_filter])
```
Comment (rafmudaf) [inline floris/simulation/turbine.py:520]: Just curious about these two lines. What's different between this and the following;
```python
self.tilt = self.tilt[duplicate_filter]
```

File: tests/turbine_unit_test.py [hunk 4]
```diff
@@ -373,4 +576,8 @@ def test_asdict(sample_inputs_fixture: SampleInputs):
     new_turb = Turbine.from_dict(dict1)
     dict2 = new_turb.as_dict()
 
+    print(dict1)
```
Comment (rafmudaf) [inline tests/turbine_unit_test.py:579]: Still needed?

**Review by rafmudaf — CHANGES_REQUESTED** [review]
I see that this pull request is still a draft, but I looked through the changes and wanted to mentioned a few things. Also, please update the examples listing [here](https://github.com/NREL/floris/blob/develop/docs/examples.md) describing the new example.

File: .gitignore (hunk 1 above)
Comment (bayc) [reply .gitignore]: Local changes, good catch! They are removed now.

File: floris/simulation/turbine.py (hunk 3 above)
Comment (bayc) [reply floris/simulation/turbine.py:520]: After talking with @RHammond2, my understanding is that this method was used when these tables were initially immutable (I modeled this after the pattern in PowerThrustTable). But sounds like the change was made along the way to have PowerThrustTable to be mutable, so I've updated to your suggestion in both tables.

File: tests/turbine_unit_test.py (hunk 4 above)
Comment (bayc) [reply tests/turbine_unit_test.py]: Nope. Removed!

**Comment by bayc** [comment]
> I see that this pull request is still a draft, but I looked through the changes and wanted to mentioned a few things. Also, please update the examples listing [here](https://github.com/NREL/floris/blob/develop/docs/examples.md) describing the new example.

Updated and added!

File: floris/simulation/farm.py [hunk 5]
```diff
@@ -25,6 +25,7 @@
     State,
     Turbine,
 )
+from floris.simulation.turbine import _compute_tilt_angles_for_floating_turbines
```
Comment (rafmudaf) [inline floris/simulation/farm.py:28]: Since `_compute_tilt_angles_for_floating_turbines` is imported outside of turbine, it should not have a prefixed underscore, or there's a reason it is meant to be private it should not be imported directly here.

File: floris/simulation/farm.py [hunk 6]
```diff
@@ -131,14 +135,27 @@ def construct_rotor_diameters(self):
     def construct_turbine_TSRs(self):
         self.TSRs = np.array([turb['TSR'] for turb in self.turbine_definitions])
 
-    def construc_turbine_pPs(self):
+    def construct_turbine_pPs(self):
         self.pPs = np.array([turb['pP'] for turb in self.turbine_definitions])
 
-    def construc_turbine_ref_density_cp_cts(self):
+    def construct_turbine_pTs(self):
+        self.pTs = np.array([turb['pT'] for turb in self.turbine_definitions])
```
Comment (rafmudaf) [inline floris/simulation/farm.py:142]: This should be listed in the attributes for this class at the top.

File: floris/simulation/farm.py [hunk 7]
```diff
@@ -131,14 +135,27 @@ def construct_rotor_diameters(self):
     def construct_turbine_TSRs(self):
         self.TSRs = np.array([turb['TSR'] for turb in self.turbine_definitions])
 
-    def construc_turbine_pPs(self):
+    def construct_turbine_pPs(self):
         self.pPs = np.array([turb['pP'] for turb in self.turbine_definitions])
```
Comment (rafmudaf) [inline floris/simulation/farm.py:139]: This should be listed in the attributes for this class at the top.

File: floris/simulation/farm.py [hunk 8]
```diff
@@ -131,14 +135,27 @@ def construct_rotor_diameters(self):
...
+    def construct_turbine_pTs(self):
+        self.pTs = np.array([turb['pT'] for turb in self.turbine_definitions])
+
+    def construct_turbine_ref_density_cp_cts(self):
         self.ref_density_cp_cts = np.array([
             turb['ref_density_cp_ct'] for turb in self.turbine_definitions
         ])
 
+    def construct_turbine_ref_tilt_cp_cts(self):
+        self.ref_tilt_cp_cts = np.array(
```
Comment (rafmudaf) [inline floris/simulation/farm.py:150]: This should be listed in the attributes for this class at the top.

File: floris/simulation/farm.py [hunk 9]
```diff
@@ -131,14 +135,27 @@ def construct_rotor_diameters(self):
...
+            [turb['ref_tilt_cp_ct'] for turb in self.turbine_definitions]
+        )
+
+    def construct_turbine_correct_cp_ct_for_tilt(self):
+        self.correct_cp_ct_for_tilt = np.array(
```
Comment (rafmudaf) [inline floris/simulation/farm.py:155]: This should be listed in the attributes for this class at the top.

File: floris/simulation/farm.py [hunk 10]
```diff
@@ -148,6 +165,9 @@ def construct_turbine_fCts(self):
     def construct_turbine_fCps(self):
         self.turbine_fCps = [(turb.turbine_type, turb.fCp_interp) for turb in self.turbine_map]
 
+    def construct_turbine_fTilts(self):
+        self.turbine_fTilts = [(turb.turbine_type, turb.fTilt_interp) for turb in self.turbine_map]
```
Comment (rafmudaf) [inline floris/simulation/farm.py:169]: This should be listed in the attributes for this class at the top.

File: floris/simulation/farm.py [hunk 11]
```diff
@@ -180,11 +200,36 @@ def expand_farm_properties(
             sorted_coord_indices,
             axis=2
         )
+        self.ref_density_cp_cts_sorted = np.take_along_axis(
```
Comment (rafmudaf) [inline floris/simulation/farm.py:203]: Need a similar method for fTilts?

**Review by rafmudaf — CHANGES_REQUESTED** [review]

File: floris/tools/floris_interface.py [hunk 12]
```diff
@@ -630,13 +658,18 @@ def get_turbine_ais(self) -> NDArrayFloat:
...
             fCt=self.floris.farm.turbine_fCts,
+            tilt_interp=self.floris.farm.turbine_fTilts,
+            correct_cp_ct_for_tilt=self.floris.farm.correct_cp_ct_for_tilt,
             turbine_type_map=self.floris.farm.turbine_type_map,
         )
         return turbine_ais
 
     @property
     def turbine_average_velocities(self) -> NDArrayFloat:
+        # TODO: Should we have a `turbine_effective_velocities` function?
```
Comment (rafmudaf) [inline floris/tools/floris_interface.py:672]: Will this be used in analysis? If so, it would be good to have.

File: floris/simulation/farm.py (hunk 5 above)
Comment (bayc) [reply floris/simulation/farm.py:28]: Updated to be a public method.

File: floris/simulation/farm.py (hunk 11 above)
Comment (bayc) [reply floris/simulation/farm.py:203]: Using the pattern from Ct/power interpolants, an array of just the different interpolators is kept (one for each kind of turbine) and the turbine type is checked to know which interpolant (fTilt) to use, so don't need to sort them.

File: floris/tools/floris_interface.py (hunk 12 above)
Comment (bayc) [reply floris/tools/floris_interface.py:672]: @rafmudaf Unsure if it would be used or not. It isn't currently, that I know of. The `turbine_average_velocities` are used in a couple of examples, but it doesn't include the effect of yaw or tilt (which would be included in the rotor effective velocity). @paulf81 What do you think?

File: floris/simulation/turbine.py [hunk 13]
```diff
@@ -125,34 +257,21 @@ def power(
     # That said, it adds overhead to convert the floats to 1-D arrays, so I don't
```
Comment (rafmudaf) [inline floris/simulation/turbine.py:257]: I think deleting the lines below this means this comment block is no longer relevant and can be deleted

File: floris/simulation/turbine.py [hunk 14]
```diff
@@ -162,7 +281,11 @@ def power(
 def Ct(
     velocities: NDArrayFloat,
     yaw_angle: NDArrayFloat,
+    tilt_angle: NDArrayFloat,
```
Comment (rafmudaf) [inline floris/simulation/turbine.py:284]: Need to add these arguments to the docstring below:
tilt_angle
ref_tilt_cp_ct
tilt_interp
correct_cp_ct_for_tilt

File: floris/simulation/turbine.py [hunk 15]
```diff
@@ -210,14 +350,18 @@ def Ct(
...
     thrust_coefficient = np.clip(thrust_coefficient, 0.0001, 0.9999)
-    effective_thrust = thrust_coefficient * cosd(yaw_angle)
+    effective_thrust = thrust_coefficient * cosd(yaw_angle) * cosd(tilt_angle - ref_tilt_cp_ct)
     return effective_thrust
 
 
 def axial_induction(
     velocities: NDArrayFloat,  # (wind directions, wind speeds, turbines, grid, grid)
     yaw_angle: NDArrayFloat,  # (wind directions, wind speeds, turbines)
+    tilt_angle: NDArrayFloat,  # (wind directions, wind speeds, turbines)
```
Comment (rafmudaf) [inline floris/simulation/turbine.py:360]: Need to add these arguments to the docstring

File: floris/simulation/turbine.py [hunk 16]
```diff
@@ -359,17 +561,22 @@ class Turbine(BaseClass):
...
+    pP: float = field()
+    pT: float = field()
+    TSR: float = field()
+    generator_efficiency: float = field()
+    ref_density_cp_ct: float = field()
+    ref_tilt_cp_ct: float = field()
     power_thrust_table: PowerThrustTable = field(converter=PowerThrustTable.from_dict)
-
-
+    # floating_tilt_table: NDArrayFloat = field(
```
Comment (rafmudaf) [inline floris/simulation/turbine.py:574]: It looks like this is initialized below so this commented code is probably out of date

File: floris/simulation/farm.py (hunk 5 above)
Comment (rafmudaf) [reply floris/simulation/farm.py:28]: I think you may have missed that change in your commit, I still see it with the underscore in the pull request

Comment (bayc) [reply floris/simulation/farm.py]: Had it locally, didn't push it up. Should be there now.

File: floris/tools/floris_interface.py [hunk 17]
```diff
@@ -642,6 +662,23 @@ def get_turbine_ais(self) -> NDArrayFloat:
     def turbine_average_velocities(self) -> NDArrayFloat:
         return average_velocity(velocities=self.floris.flow_field.u)
 
+    @property
+    def turbine_effective_velocities(self) -> NDArrayFloat:
```
Comment (bayc) [inline floris/tools/floris_interface.py:666]: @rafmudaf Added `turbine_effective_velocities` at the `floris_interface` level. Should this be an `@property`? Modelled after `turbine_average_velocities`.

**Comment by paulf81** [comment]
The latest commit [adding turbine_effective_velocities to floris_interface](https://github.com/NREL/floris/pull/518/commits/138f56a74367da0325aed6ed5f6470f0668765a3) I think is causing example 24 to break,

**Comment by bayc** [comment]
> The latest commit [adding turbine_effective_velocities to floris_interface](https://github.com/NREL/floris/pull/518/commits/138f56a74367da0325aed6ed5f6470f0668765a3) I think is causing example 24 to break,

Thanks @paulf81, I've updated so that all the examples work now.

**Comment by paulf81** [comment]
Hi @bayc , I was thinking I was going to switch the example to use an external library folder, instead of the !include syntax to reference the floating turbine model, but I think this path would have to be specified so then it seems better actually to leave these as is for now, and maybe we can include syntax on using the external library folder to documentation rather than examples, what do you think @rafmudaf ?

**Review by rafmudaf — APPROVED** [review]
